frame/lockdrop/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This file is part of Kulupu.
+//
+// Copyright (c) 2020 Wei Tang.
+//
+// Kulupu is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Kulupu is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Kulupu. If not, see <http://www.gnu.org/licenses/>.
+
+//! Era information recording.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Encode, Decode};
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+use sp_std::{cmp, prelude::*};
+use sp_runtime::RuntimeDebug;
+use frame_support::{
+    ensure, decl_storage, decl_module, decl_event, decl_error, 
+    traits::{Currency, LockableCurrency, WithdrawReasons},
+};
+use frame_system::{ensure_root, ensure_signed};
+
+pub type CampaignIdentifier = [u8; 4];
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct CampaignInfo<T: Config> {
+    end_block: T::BlockNumber,
+    min_lock_end_block: T::BlockNumber,
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct LockInfo<T: Config> {
+    end_block: T::BlockNumber,
```

Comment:
i think to make this pallet more helpful, also holding the amount locked is a good idea.